src/python-common/ceph/cryptotools/cryptotools.py

```diff
@@ -0,0 +1,135 @@
+"""
+This file has been isolated into a module so that it can be run
+in a subprocess therefore sidestepping the
+`PyO3 modules may only be initialized once per interpreter process` problem.
+"""
+
+from typing import Any, Dict
+
+import argparse
+import json
+import sys
+
+from argparse import Namespace
+
+from .internal import InternalCryptoCaller, InternalError
+
+
+def _read() -> str:
+    return sys.stdin.read()
+
+
+def _load() -> Dict[str, Any]:
+    return json.loads(_read())
+
+
+def _respond(data: Dict[str, Any]) -> None:
+    json.dump(data, sys.stdout)
+
+
+def _write(content: str) -> None:
+    sys.stdout.write(content)
+    sys.stdout.flush()
+
+
+def _fail(msg: str, code: int = 0) -> Any:
+    json.dump({'error': msg}, sys.stdout)
+    sys.exit(code)
+
+
+def password_hash(args: Namespace) -> None:
+    data = _load()
+    password = data['password']
+    salt_password = data['salt_password']
+    hash_str = args.crypto.password_hash(password, salt_password)
+    _respond({'hash': hash_str})
+
+
+def verify_password(args: Namespace) -> None:
+    data = _load()
+    password = data.get('password', '')
+    hashed_password = data.get('hashed_password', '')
+    try:
+        ok = args.crypto.verify_password(password, hashed_password)
+    except ValueError as err:
+        _fail(str(err))
+    _respond({'ok': ok})
+
+
+def create_private_key(args: Namespace) -> None:
+    _write(args.crypto.create_private_key())
+
+
+def create_self_signed_cert(args: Namespace) -> None:
+    data = _load()
+    dname = data['dname']
+    private_key = data['private_key']
+    _write(args.crypto.create_self_signed_cert(dname, private_key))
+
+
+def certificate_days_to_expire(args: Namespace) -> None:
+    crt = _read()
+    try:
+        days_until_exp = args.crypto.certificate_days_to_expire(crt)
+    except InternalError as err:
+        _fail(str(err))
+    _respond({'days_until_expiration': days_until_exp})
+
+
+def get_cert_issuer_info(args: Namespace) -> None:
+    crt = _read()
+    org_name, cn = args.crypto.get_cert_issuer_info(crt)
+    _respond({'org_name': org_name, 'cn': cn})
+
+
+def verify_tls(args: Namespace) -> None:
+    data = _load()
+    crt = data['crt']
+    key = data['key']
+    try:
+        args.crypto.verify_tls(crt, key)
+    except ValueError as err:
+        _fail(str(err))
+    _respond({'ok': True})  # need to emit something on success
+
+
+def main() -> None:
+    # create the top-level parser
```

Comment:
I wonder if there is a generic mechanism for this. It has to be somewhat common.

Reply:
I'm not sure what you are referring to here, setting up an argument parser?

Reply:
I mean a python module that is providing RPCs over a socket / pipe.